src/veins/modules/utility/TimerManager.cc

```diff
@@ -0,0 +1,167 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+//
+#include "veins/modules/utility/TimerManager.h"
+
+using omnetpp::simTime;
+using omnetpp::simtime_t;
+using Veins::TimerManager;
+using Veins::TimerSpecification;
+
+TimerSpecification::TimerSpecification(std::function<void()> callback)
+	: start_mode_(StartMode::IMMEDIATE), end_mode_(EndMode::OPEN), period_(-1), callback_(callback) {}
+
+TimerSpecification &TimerSpecification::setInterval(simtime_t interval) {
+  ASSERT(interval > 0);
+  period_ = interval;
+  return *this;
+}
+
+TimerSpecification &TimerSpecification::setRelativeStart(simtime_t start) {
+  start_mode_ = StartMode::RELATIVE;
+  start_ = start;
+  return *this;
+}
+
+TimerSpecification &TimerSpecification::setAbsoluteStart(simtime_t start) {
+  start_mode_ = StartMode::ABSOLUTE;
+  start_ = start;
+  return *this;
+}
+
+TimerSpecification &TimerSpecification::setImmediateStart() {
+  start_mode_ = StartMode::IMMEDIATE;
+  return *this;
+}
+
+TimerSpecification &TimerSpecification::setRelativeEnd(simtime_t end) {
+  end_mode_ = EndMode::RELATIVE;
+  end_time_ = end;
+  return *this;
+}
+
+TimerSpecification &TimerSpecification::setAbsoluteEnd(simtime_t end) {
+  end_mode_ = EndMode::ABSOLUTE;
+  end_time_ = end;
+  return *this;
+}
+
+TimerSpecification &TimerSpecification::setRepititions(size_t n) {
+  end_mode_ = EndMode::REPITITION;
+  end_count_ = n;
+  return *this;
+}
+
+TimerSpecification &TimerSpecification::setOpenEnd() {
+  end_mode_ = EndMode::OPEN;
+  return *this;
+}
+
+TimerSpecification &TimerSpecification::setOneshotIn(omnetpp::simtime_t in) {
+	return this->setRelativeStart(in).setInterval(1).setRepititions(1);
+}
+
+TimerSpecification &TimerSpecification::setOneshotAt(omnetpp::simtime_t at) {
+  return this->setAbsoluteStart(at).setInterval(1).setRepititions(1);
+}
+
+void TimerSpecification::finalize() {
+  switch (start_mode_) {
+  case StartMode::RELATIVE:
+    start_ += simTime();
+    start_mode_ = StartMode::ABSOLUTE;
+    break;
+  case StartMode::ABSOLUTE:
+    break;
+  case StartMode::IMMEDIATE:
+    start_ = simTime() + period_;
+    break;
+  }
+
+  switch (end_mode_) {
+  case EndMode::RELATIVE:
+    end_time_ += simTime();
+    end_mode_ = EndMode::ABSOLUTE;
+    break;
+  case EndMode::ABSOLUTE:
+    break;
+  case EndMode::REPITITION:
+    end_time_ = start_ + ((end_count_ - 1) * period_);
+    end_mode_ = EndMode::ABSOLUTE;
+  case EndMode::OPEN:
+    break;
+  }
+}
+
+bool TimerSpecification::validOccurence(simtime_t time) const {
+	return time >= start_ && (time <= end_time_ || end_mode_ == EndMode::OPEN) && omnetpp::fmod(time - start_, period_) == 0;
```

Comment:
Split up parts of condtion into separate bools for increased readibility.
Indentation error (Tab here)